crates/ursa-network/src/behaviour.rs

```diff
@@ -301,6 +279,80 @@ impl<P: StoreParams> Behaviour<P> {
     pub fn unsubscribe(&mut self, topic: &Topic) -> Result<bool, PublishError> {
         self.gossipsub.unsubscribe(topic)
     }
+}
+
+/// A `Networkbehaviour` that handles Ursa's different protocol implementations.
+///
+/// The poll function must have the same signature as the NetworkBehaviour
+/// function and will be called last within the generated NetworkBehaviour implementation.
+///
+/// The events generated [`BehaviourEvent`].
+pub struct Behaviour<P: StoreParams> {
+    inner: InternalBehaviour<P>,
+
+    /// Ursa's emitted events.
+    events: VecDeque<BehaviourEvent>,
+
+    /// Pending requests
+    pending_requests: HashMap<RequestId, ResponseChannel<UrsaExchangeResponse>>,
+
+    /// Pending responses
+    pending_responses: HashMap<RequestId, oneshot::Sender<Result<UrsaExchangeResponse>>>,
+
+    queries: FnvHashMap<QueryId, BitswapInfo>,
+}
+
+impl<P: StoreParams> Behaviour<P> {
+    pub fn new<S: BitswapStore<Params = P>>(
+        keypair: &Keypair,
+        config: &NetworkConfig,
+        bitswap_store: S,
+        relay_client: Option<libp2p::relay::v2::client::Client>,
+    ) -> Self {
+        Self {
+            inner: InternalBehaviour::new(keypair, config, bitswap_store, relay_client),
+            events: VecDeque::new(),
+            pending_requests: HashMap::default(),
+            pending_responses: HashMap::default(),
+            queries: Default::default(),
+        }
+    }
+
+    pub fn publish(
+        &mut self,
+        topic: Topic,
+        data: GossipsubMessage,
+    ) -> Result<MessageId, PublishError> {
+        self.inner.publish(topic, data)
+    }
+
+    pub fn public_address(&self) -> Option<&Multiaddr> {
+        self.inner.public_address()
+    }
+
+    pub fn peers(&self) -> HashSet<PeerId> {
+        self.inner.peers()
+    }
+
+    pub fn is_relay_client_enabled(&self) -> bool {
+        self.inner.is_relay_client_enabled()
+    }
+
+    pub fn discovery(&mut self) -> &mut DiscoveryBehaviour {
+        self.inner.discovery()
+    }
+
+    pub fn bootstrap(&mut self) -> Result<kad::QueryId, Error> {
+        self.inner.bootstrap()
+    }
+
+    pub fn subscribe(&mut self, topic: &Topic) -> Result<bool, SubscriptionError> {
+        self.inner.subscribe(topic)
+    }
+
+    pub fn unsubscribe(&mut self, topic: &Topic) -> Result<bool, PublishError> {
+        self.inner.unsubscribe(topic)
+    }
 
     pub fn publish_ad(&mut self, public_address: Multiaddr) -> Result<()> {
```

Comment:
all other events.push_back were replaced, did you miss this one or is it intentional?

Reply:
Yes, I did. Will fix